subsys/logging/log_core.c

```diff
@@ -510,51 +509,49 @@ u32_t log_filter_get(struct log_backend const *const backend,
 
 char *log_strdup(const char *str)
 {
-	u32_t *dupl;
-	char *sdupl;
+	struct log_strdup_buf *dup;
 	int err;
 
-	err = k_mem_slab_alloc(&log_strdup_pool, (void **)&dupl, K_NO_WAIT);
+	err = k_mem_slab_alloc(&log_strdup_pool, (void **)&dup, K_NO_WAIT);
 	if (err) {
 		/* failed to allocate */
 		return (char *)log_strdup_fail_msg;
 	}
 
 	/* Set 'allocated' flag. */
-	*dupl = 1;
-	dupl++;
-	sdupl = (char *)dupl;
+	atomic_set(&dup->refcount, 1);
 
-	strncpy(sdupl, str, CONFIG_LOG_STRDUP_MAX_STRING - 1);
-	sdupl[LOG_STRBUF_STR_SIZE - 1] = '\0';
-	sdupl[LOG_STRBUF_STR_SIZE - 2] = '~';
+	strncpy(dup->buf, str, sizeof(dup->buf) - 2);
+	dup->buf[sizeof(dup->buf) - 2] = '~';
+	dup->buf[sizeof(dup->buf) - 1] = '\0';
 
-	return sdupl;
+	return dup->buf;
 }
 
 bool log_is_strdup(void *buf)
 {
+	struct log_strdup_buf *dup;
+
 	/* Lowest possible address is located at the second word of the first
 	 * buffer in the pool. First word is dedicated for 'allocated' flag.
 	 *
 	 * Highest possible address is the second word of the last buffer in the
 	 * pool.
 	 */
-	static const void *start = log_strdup_pool_buf + sizeof(u32_t);
+	static const void *start = log_strdup_pool_buf + sizeof(dup->refcount);
 	static const void *end = &log_strdup_pool_buf[
 						 LOG_STRDUP_POOL_BUFFER_SIZE
-					       - LOG_STRBUF_BUF_SIZE
-					       + sizeof(u32_t)];
+					       - sizeof(dup->buf)];
 	return (buf >= start) && (buf <= end);
 }
 
 void log_free(void *str)
 {
-	u32_t *buf = (u32_t *)str;
+	struct log_strdup_buf *dup = (struct log_strdup_buf *)
```

Comment:
use CONTAINER_OF() to get structure address.

Reply:
done